rust/ql/lib/codeql/rust/internal/Type.qll

```diff
@@ -140,21 +128,14 @@ class UnitType extends TupleType {
   override string toString() { result = "()" }
 }
 
-abstract private class StructOrEnumType extends Type {
-  abstract ItemNode asItemNode();
-}
-
 /** A struct type. */
-class StructType extends StructOrEnumType, TStruct {
+class StructType extends Type, TStruct {
   private Struct struct;
 
   StructType() { this = TStruct(struct) }
 
-  override ItemNode asItemNode() { result = struct }
-
-  override StructField getStructField(string name) { result = struct.getStructField(name) }
-
-  override TupleField getTupleField(int i) { result = struct.getTupleField(i) }
+  /** Get the struct that this struct type represents. */
+  Struct asStruct() { result = struct }
```

Comment:
nit: I think `getStruct` is better; `as` is normally used when there is a possibility that the predicate does not hold.

Reply:
Alternatively, move the predicate all the way up into `Type`; then you can replace all of the `(StructType).asStruct()` occurrences with `.asStruct()`.